File: semana 6/decorator.cpp
#include <iostream>
#include <memory>
#include <cmath>
using namespace std;

// Componente base
class FuncionMatematica {
public:
    virtual double calcular(double x) = 0;
    virtual ~FuncionMatematica() = default;
};

// Componente concreto
class RaizCuadrada : public FuncionMatematica {
public:
    double calcular(double x) override {
        return sqrt(x);
    }
};

// Decorator base
class ValidadorDecorator : public FuncionMatematica {
protected:
    shared_ptr<FuncionMatematica> funcion;

public:
    ValidadorDecorator(shared_ptr<FuncionMatematica> f)
        : funcion(f) {}

    double calcular(double x) override {
        return funcion->calcular(x);
    }
};

// Decorador concreto: validación
class ValidadorPositivos : public ValidadorDecorator {
public:
    ValidadorPositivos(shared_ptr<FuncionMatematica> f)
        : ValidadorDecorator(f) {}

    double calcular(double x) override {
        if (x <= 0) {
            throw invalid_argument("La funcion requiere valores positivos");
        }

        cout << "[Validador] Entrada valida: " << x << endl;
        return funcion->calcular(x);
    }
};

// Decorador concreto: logging
class LoggerDecorator : public ValidadorDecorator {
public:
    LoggerDecorator(shared_ptr<FuncionMatematica> f)
        : ValidadorDecorator(f) {}

    double calcular(double x) override {
        cout << "[Logger] Calculando para x = " << x << endl;
        double resultado = funcion->calcular(x);
        cout << "[Logger] Resultado: " << resultado << endl;
        return resultado;
    }
};

int main() {
    cout << "=== Decorator Pattern ===" << endl;

    try {
        // Componente base
        auto raiz = make_shared<RaizCuadrada>();

        // Decorar con validación
        auto validada = make_shared<ValidadorPositivos>(raiz);

        // Decorar con logging
        auto con_log = make_shared<LoggerDecorator>(validada);

        cout << "\nPrueba con valor valido (25):" << endl;
        cout << "Resultado: " << con_log->calcular(25) << endl;

        cout << "\nPrueba con valor invalido (-4):" << endl;
        cout << "Resultado: " << con_log->calcular(-4) << endl;
    }
    catch (const exception& e) {
        cerr << "Error: " << e.what() << endl;
    }

    return 0;
}

File: semana 6/decorator.py
import math
from functools import wraps


def validar_positivos(func):
    """Decorator que valida entrada positiva"""

    @wraps(func)  # Preserva metadatos de la función
    def wrapper(*args, **kwargs):
        if any(x <= 0 for x in args):
            raise ValueError("La función requiere valores positivos")

        print(f"[Validador] Entrada válida: {args[0]}")
        return func(*args, **kwargs)

    return wrapper


def logger(func):
    """Decorator que añade logging"""

    @wraps(func)
    def wrapper(*args, **kwargs):
        print(f"[Logger] Calculando para x = {args[0]}")
        resultado = func(*args, **kwargs)
        print(f"[Logger] Resultado: {resultado}")
        return resultado

    return wrapper


# Aplicar múltiples decoradores
@logger
@validar_positivos
def raiz_cuadrada(n: float) -> float:
    return math.sqrt(n)


if __name__ == "__main__":
    print("=== Decorator Pattern ===")

    print("\nPrueba con valor válido (25):")
    print(f"Resultado: {raiz_cuadrada(25)}")

    print("\nPrueba con valor inválido (-4):")
    try:
        print(f"Resultado: {raiz_cuadrada(-4)}")
    except ValueError as e:
        print(f"Error: {e}")

    print(f"\n[Nota] @wraps preserva: {raiz_cuadrada.__name__}")
